synapse/storage/databases/main/event_push_actions.py

```diff
@@ -436,42 +454,70 @@ def _get_unread_counts_by_pos_txn(
             """,
             (room_id, user_id, receipt_stream_ordering, receipt_stream_ordering),
         )
-        row = txn.fetchone()
+        max_summary_stream_ordering = 0
+        for summary_stream_ordering, notif_count, unread_count, thread_id in txn:
+            if thread_id == "main":
+                counts = NotifCounts(
+                    notify_count=notif_count, unread_count=unread_count
+                )
+            # TODO Delete zeroed out threads completely from the database.
+            elif notif_count or unread_count:
+                thread_counts[thread_id] = NotifCounts(
+                    notify_count=notif_count, unread_count=unread_count
+                )
```

Comment:
Copied from https://github.com/matrix-org/synapse/pull/13181#discussion_r955193227:

> Might be worth looking at this again briefly not that some other bugs have been fixed.

Reply:
It seems silly to delete rows that are likely to come back into play "soon", so I made two changes:

1. I updated the SQL query to ignore these rows (instead of returning them from the database and ignoring them in Python).
2. I added some code to delete empty push summaries that are over a month old. Those seem like rows that won't be needed again soon (and they can always be recreated).

Reply:
> It seems silly to delete rows that are likely to come back into play "soon"

It's worth noting that an `UPDATE` is often more a less done via a `DELETE + INSERT` in the guts of postgres, so doing `UPDATE + UPDATE` is not obviously a win over `DELETE + INSERT`. Not sure if that changes anything?

Reply:
Though actually HOT updates come into play here a bit, most likely: https://www.cybertec-postgresql.com/en/hot-updates-in-postgresql-for-better-performance/

Reply:
> > It seems silly to delete rows that are likely to come back into play "soon"
> 
> It's worth noting that an `UPDATE` is often more a less done via a `DELETE + INSERT` in the guts of postgres, so doing `UPDATE + UPDATE` is not obviously a win over `DELETE + INSERT`. Not sure if that changes anything?

Maybe I'm over thinking, but I guess it lets the database engine potentially be smarter? 🤷

Reply:
From discussion it seems that this is likely a bit of premature optimization, we're going to just file an issue for now.

Reply:
See #13875.